### PR TITLE
Music looping

### PR DESCRIPTION
**Pull request by cadenceblorbo**
- Extends MIDI2EventUnity to support splitting a song up into multiple components for looping
- Provides a way to dynamically advance past song loops
- Creates an AudioSystem to manage several MIDI2EventUnity songs and fade between them using an Audio Mixer Controller
- Makes pausing audio a lot simpler (menu sounds can use ignoreListenerPause to play in the menu)

Check out the MusicRelated/AudioSysTests scene for a demo!

### CODE REVIEW

**Comment by rping-qwfpgj** [comment]
:+1:

File: Assets/Scripts/GameManagement/GameStates/PauseState.cs
Comment (wormsbeeeee) [inline Assets/Scripts/GameManagement/GameStates/PauseState.cs:1]: Nice change to this, makes it a lot cleaner

File: Assets/Scripts/MIDI2Event/AudioSystem.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        {
+            float volStep = Mathf.Lerp(startingVol, fadedDB, elapsedTime / fadeTime);
+            Mixer.SetFloat(previous.VolumeSliderName, volStep);
+            elapsedTime += Time.deltaTime;
+            yield return null;
+        }
+        previous.Stop();
+        StartNextIfRelevant(next);
+        Mixer.SetFloat(previous.VolumeSliderName, startingVol);
+    }
```
Comment (wormsbeeeee) [inline Assets/Scripts/MIDI2Event/AudioSystem.cs:75]: Fading seems to work really well in the test scene when going back and forth between the two songs

**Review by wormsbeeeee — CHANGES_REQUESTED** [review]
Overall this is a fantastic system and the track looping works really well, I can't even tell where it's looping unless I'm really focused on listening for it. MIDI2EventUnity as a class is really helpful for seeing stuff in the Editor too. Advancing the track works super well too to break out of the loop.

When playing the test scene, I went to play Breaking Out first, which worked perfectly and looped perfectly - then I went to look at the other one, which I believe is playing "testaudio". It fades into it amazingly too, and plays fine, I just noticed that it wasn't looping, which I believe is intentional since it doesn't have a loop midi file right? And it's obviously not set to loop in the inspector either, so this isn't a problem. **Some things** I noticed:

- Entering the scene and immediately switching off of Breaking Out to testaudio fades out Breaking Out and fades in testaudio, but Breaking Out will start playing after about 70 seconds-ish while testaudio is still playing. 
  - running over to the green advance button **immediately** after pressing the testaudio button will cause the very end of Breaking Out to play immediately instead. I assume this is because Breaking Out isn't completely "finished" playing by the time I hit the button, since it has to fade out, and hitting the advance button so quickly triggers it to play the end instead of continuing to fade out.
  - worth noting that Breaking Out will start playing from what I assume is its loop point, not from the start of the song. 
  - testaudio will finish normally and Breaking Out will continue to play **and loop** accordingly. The advance button seems to still work at stopping the loop too after this point.
  - this behavior seems to happen when switching off of Breaking Out at any point before it reaches the 70-ish second mark (where I'm guessing the loop begins). Basically if you wait 30 seconds and then switch over to testaudio, it will still happen, just 40 seconds later instead.

This seems to happen when loading into the scene and running over to switch songs before Breaking Out's loop, since after waiting long enough after starting the scene before switching doesn't cause any issues. It's almost like switching songs before it reaches the loop doesn't "stop" Breaking Out but "mutes" it instead, and then starts playing from the loop at regular volume again. 

Regardless, the rest of the system seems remarkably well-written. I noticed that the moving platforms and pendulum will always stop at a "set" position - I'm not sure how to say that but what I mean is that the moving platform won't stop in the middle of its path when the song stops or switches and that's great. Nice job!!

**Comment by cadenceblorbo** [comment]
> Overall this is a fantastic system and the track looping works really well, I can't even tell where it's looping unless I'm really focused on listening for it. MIDI2EventUnity as a class is really helpful for seeing stuff in the Editor too. Advancing the track works super well too to break out of the loop.
> 
> When playing the test scene, I went to play Breaking Out first, which worked perfectly and looped perfectly - then I went to look at the other one, which I believe is playing "testaudio". It fades into it amazingly too, and plays fine, I just noticed that it wasn't looping, which I believe is intentional since it doesn't have a loop midi file right? And it's obviously not set to loop in the inspector either, so this isn't a problem. **Some things** I noticed:
> 
>     * Entering the scene and immediately switching off of Breaking Out to testaudio fades out Breaking Out and fades in testaudio, but Breaking Out will start playing after about 70 seconds-ish while testaudio is still playing.
>       
>       * running over to the green advance button **immediately** after pressing the testaudio button will cause the very end of Breaking Out to play immediately instead. I assume this is because Breaking Out isn't completely "finished" playing by the time I hit the button, since it has to fade out, and hitting the advance button so quickly triggers it to play the end instead of continuing to fade out.
>       * worth noting that Breaking Out will start playing from what I assume is its loop point, not from the start of the song.
>       * testaudio will finish normally and Breaking Out will continue to play **and loop** accordingly. The advance button seems to still work at stopping the loop too after this point.
>       * this behavior seems to happen when switching off of Breaking Out at any point before it reaches the 70-ish second mark (where I'm guessing the loop begins). Basically if you wait 30 seconds and then switch over to testaudio, it will still happen, just 40 seconds later instead.
> 
> 
> This seems to happen when loading into the scene and running over to switch songs before Breaking Out's loop, since after waiting long enough after starting the scene before switching doesn't cause any issues. It's almost like switching songs before it reaches the loop doesn't "stop" Breaking Out but "mutes" it instead, and then starts playing from the loop at regular volume again.
> 
> Regardless, the rest of the system seems remarkably well-written. I noticed that the moving platforms and pendulum will always stop at a "set" position - I'm not sure how to say that but what I mean is that the moving platform won't stop in the middle of its path when the song stops or switches and that's great. Nice job!!

Great catch! These issues should be fixed in the most current version, I was forgetting to cancel all scheduling when calling Stop.

**Pull request merged**